src/sp/it/pl/util/file/Util.kt

```diff
@@ -13,53 +13,49 @@ import kotlin.streams.asStream
 import kotlin.streams.toList
 
 /**
- * For files 'filename.extension' is returned.
- * For directories only name is returned.
- * Root directory returns 'X:\' string.
- *
- * Use instead of [File.getName] which returns empty string for root directories.
+ * Returns the [File.getName] or [File.toString] if this is the root directory,
+ * since that returns an empty string otherwise.
  *
  * @return name of the file with suffix
  */
-val File.nameOrRoot: String get() = name.takeUnless { it.isEmpty() } ?: toString()
+val File.nameOrRoot: String
+    get() = name.takeUnless { it.isEmpty() } ?: toString()
 
 /**
- * For files name with no extension is returned.
- * For directories name is returned.
- * Root directory returns 'X:\' string.
+ * Returns the [File.nameWithoutExtension] or [File.nameOrRoot] if this is a directory.
  *
- * @return name of the file without suffix
+ * @return name of the file without extension
  */
-// TODO: does not work for directories with '.'
-val File.nameWithoutExtensionOrRoot: String get() = nameWithoutExtension.takeUnless { it.isEmpty() } ?: toString()
+val File.nameWithoutExtensionOrRoot: String
+    get() = if(isDirectory) nameWithoutExtension else nameOrRoot
 
-/** @returns file itself if exists or its first existing parent or error if null or no parent exists */
+/** @return file itself if exists or its first existing parent or error if no parent exists */
 fun File.find1stExistingParentFile(): Try<File, Void> = when {
     exists() -> Try.ok(this)
     else -> parentDir?.find1stExistingParentFile() ?: Try.error()
 }
 
-/** @returns file's first existing parent or error if null or no parent exists */
+/** @return first existing directory in this file's hierarchy or error if no parent exists */
 fun File.find1stExistingParentDir(): Try<File, Void> = when {
     exists() && isDirectory -> Try.ok(this)
     else -> parentDir?.find1stExistingParentDir() ?: Try.error()
 }
 
-fun File.childOf(childName: String) = File(this, childName)
+inline fun File.childOf(childName: String) = File(this, childName)
```

Comment:
they are because they are just shortcuts for existing methods and you might as well remove the unneccessary call overhead

Reply:
There is no call overhead because the inlining will happen anyway due to Hotspot optimizations (https://stackoverflow.com/questions/1159087/inlining-in-java).

Inline functions are useful not because they make inlining possible (inlining is already happening everywhere automatically), but because they make it possible to remove unnecessary lambda allocations (https://kotlinlang.org/docs/reference/inline-functions.html), which Hotspot itself would not be able to do -> the lambdas will not even make it past the compiler.

Right here, Idea itself should be showing you a warning that inlining thiese methods carries no performance improvements and should be avoided. Inlining should only be done if thers lambdas or generics involved and only if there is something to be gained at all. This is not such case. Pls revert inlining.

Reply:
That's why I suppress the nothing to inline warning at the top of the file, it is useless imo. It also shaves off a bit of size actually, since the function consumes is longer than the actual call. Why would we not inline them upfront? I know that it may be unnecessary, but I can't imagine a case where it would be harmful here.

Reply:
That file scoped suppress totally escaped me. Pls remove that as well.

So you received a warning from the IDE and instead of going by it you decided to suppress it for no good reason despite it clearly saying that what you did is not good idea. Why do you think this warning is useless? Because it warns you against use of a feature that will never bring harm so we better use it more than less? Unlike Java, Kotlin suggestions/warnings are almost always right.

Have you read the link I sent about Kotlin inline feature?

Reasons against:
- theres no lambdas to inline and no wasteful object allocation to fix
- theres JIT inlining which will take care of this so there's no performance to be gained
- Kotlin inlining is for lambda/reified generics, using it outside of that causes confusion, complexity, maintenance and readability overheads
- Compiled code is larger

There is no benefit and plenty detriments here and the simplest design wins.  It was a bad idea.

Reply:
> - theres no lambdas to inline and no wasteful object allocation to fix
> - theres JIT inlining which will take care of this so there's no performance to be gained

These are not reasons for not using it, these are reasons why it might be unnecessary.

> - Kotlin inlining is for lambda/reified generics, using it outside of that causes confusion, complexity, maintenance and readability overheads

Debatable, I for my part do not think it's less readable this way

> - Compiled code is larger

No it isn't, because as I stated, the method calls are actually longer than their bodies.

Reply:
Okay, I'll revert it.

 But in one point I still have to disagree: The compiled code won't be larger in this case, since `file.childOf(childName)` is the same amount of statements as `File(file, childName)`, namely 1.

Reply:
You are probably right about that. But as with all optimizations - dev should not make guesses and assumptions and let the tools do the job. If they can't do it sufficiently well, then the dev should take care of it..